src/qibolab/instruments/oscillator.py

```diff
@@ -19,40 +47,138 @@ class LocalOscillator(Instrument):
-        if power is not None:
-            self.power = power
-        if frequency is not None:
-            self.frequency = frequency
+        If the new value is the same with the cached value, it is not updated.
+        """
+        if self.power != x:
+            self.settings.power = x
+            if self.is_connected:
+                self.device.set("power", x)
+
+    @property
+    def reference_clock_source(self):
+        return self._reference_clock_source
+
+    @reference_clock_source.setter
+    def reference_clock_source(self, x):
+        """Switch the reference clock source of the local oscillator.
+
+        The value is cached in the :class:`qibolab.instruments.oscillator.LocalOscillator`
+        class. If we are connected to the instrument when the setter is called, it is also
+        automatically uploaded to the instruments. If we are not connected the cached value
+        is automatically uploaded when we connect.
+        """
+        self._reference_clock_source = x
+        if self.is_connected:
+            self.device.set("ref_osc_source", x)
+
+    def create(self):
+        """Create instance of physical device."""
+        return DummyDevice()
+
+    def connect(self):
+        """Connects to the instrument using the IP address set in the runcard."""
+        if not self.is_connected:
+            for attempt in range(self.RECONNECTION_ATTEMPTS):
+                try:
+                    self.device = self.create()
+                    self.is_connected = True
+                    break
+                except KeyError as exc:
+                    log.info(f"Unable to connect:\n{str(exc)}\nRetrying...")
+                    self.name += "_" + str(attempt)
+                except ConnectionError as exc:
+                    log.info(f"Unable to connect:\n{str(exc)}\nRetrying...")
+            if not self.is_connected:
+                raise InstrumentException(self, f"Unable to connect to {self.name}")
+        else:
+            raise InstrumentException(self, "There is an open connection to the instrument already")
+        self.upload()
+
+    def upload(self):
+        """Uploads cached setting values to the instruments."""
+        if not self.is_connected:
+            raise InstrumentException(self, "Cannot upload settings if instrument is not connected.")
+
+        if self.settings.frequency is not None:
+            self.device.set("frequency", self.settings.frequency)
+        else:
+            self.settings.frequency = self.device.get("frequency")
+
+        if self.settings.power is not None:
+            self.device.set("power", self.settings.power)
+        else:
+            self.settings.power = self.device.get("power")
+
+        if self.reference_clock_source is not None:
+            self.reference_clock_source = self._reference_clock_source
+        else:
+            self._reference_clock_source = self.device.get("ref_osc_source")
```

Comment:
Here there is definitely a pattern, and I believe it is not even limited to local oscillators (but I might be wrong).

Couldn't we make a helper function in `Instrument` for this?
Something like:
```py
def update(self, name: str, dname: Optional[str] = None):
    if dname is None:
        dname = name

    value = getattr(self.settings, name)
    if value is not None:
        self.device.set(dname, value)
    else:
        setattr(self.settings, name, self.device.get(dname))
```

Reply:
Done, I renamed the reference clock to `ref_osc_source` so that we don't need the `dname` distinction and also moved it to `settings` to simplify.